[#5200]Update the Column data type for python client

SophieTech88 opened the pull request:
### What changes were proposed in this pull request?
1. Implement type.java as type.py in Python client
2.  Implement types.java as types.py in Python client
3.  Create a test_types.py in unit tests.

### Why are the changes needed?
We need to support the column data type in python client

Fix: #5200

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Need to add the unit tests

jerryshao commented:
@mchades can you please help to review?

jerryshao commented:
@noidname01 , would you please also help to review this PR? Thanks a lot.

SophieTech88 commented:
When I double check the types folder (gravitino/tree/main/api/src/main/java/org/apache/gravitino/rel/types/) I find it has another file Decimal.java, so just update a new commit to convert decimal.java to decimal.python with unit test in python client folder.

mchades commented on clients/client-python/gravitino/api/decimal.py:
missing license header

SophieTech88 replied:
Many thanks. Just added the license header.

mchades commented on clients/client-python/gravitino/api/decimal.py:
```diff
@@ -0,0 +1,59 @@
+from .types import Types
+
+
+class Decimal:
```
Here is the reason why we need to wrap Java `BigDecimal`:
https://github.com/apache/gravitino/blob/6d05ec4b8bf278c2dd7a36af35589f07b220fcab/api/src/main/java/org/apache/gravitino/rel/types/Decimal.java#L27-L34

Does the `PyDecimal` have the same issue?

SophieTech88 replied:
Thank you so much to point out the reason why we need to wrap Java BigDecimal. Then I just double checked for Python PyDecimal, it tends to handle this more intuitively for currency by counting all visible digits. So for example:
```
from decimal import Decimal

value = Decimal("0.01") ## it is 2
print("Precision:", value.adjusted() + 1)  ## it is 2 
print("Scale:", -value.as_tuple().exponent)
```
So PyDecimal didn't have the similar issue like Java BigDecimal. 
I will remove the new change for decimal.py.

SophieTech88 replied:
Removed. Done.

mchades commented on clients/client-python/tests/unittests/test_decimal.py:
miss license header

mchades commented on clients/client-python/gravitino/api/types.py:
```diff
@@ -0,0 +1,921 @@
+    class NullType(Type):
+        _instance = None
+
+        @classmethod
+        def get(cls):
+            if cls._instance is None:
+                cls._instance = cls()
+            return cls._instance
```
I'm not sure if these codes can guarantee the singleton of `_instance` under race conditions in Python, can you help confirm it? @xunliu @noidname01

SophieTech88 replied:
Thank you for catching that! You’re absolutely correct—this implementation does not currently handle race conditions, and under high concurrency, there’s a chance multiple threads could attempt to create the singleton instance simultaneously, breaking the singleton pattern.

Solution: Adding Thread Safety with a Lock
To handle this, we can use threading.Lock, which ensures that only one thread can enter the critical section where the instance is created. 

Here is the example:
```
class template:
    _instance = None
    _lock = threading.Lock()  # A lock to ensure thread safety

    @classmethod
    def get(cls):
        with cls._lock:  # Ensures only one thread can access this block at a time
            if cls._instance is None:
                cls._instance = cls()
        return cls._instance
 ```

mchades replied:
`Lock` makes things more complex, is there another simple common method?

SophieTech88 replied:
Another option is use __new__():
```
class Types:

    class NullType(Type):
        _instance = None

        @classmethod
        def get(cls):
            if cls._instance is None:
                cls._instance = cls.__new__(cls)
            return cls._instance

```
logic here:
- get() checks if _instance is None.
- If _instance is None, cls.__new__(cls) is called to create a new instance of NullType without directly calling the constructor __init__.
- get() then returns the instance. Subsequent calls to get() will return the existing _instance.

jerryshao replied:
Does Python support singleton?

SophieTech88 replied:
Yes, Python does support the Singleton pattern, which ensures that only one instance of a class is created. While Python does not enforce singletons natively, you can implement it in several ways.

xunliu reviewed (commented):
hi @SophieTech88 
I help you improve this PR, please help me review my change parts.
1. We can use `__new__()` implement singleton.
2. Added some return type and function param type.
3. Added some class and function comments.

mchades commented on clients/client-python/gravitino/api/types.py:
```diff
@@ -0,0 +1,1106 @@
+
+        @classmethod
+        def get(cls) -> "BooleanType":
+            return cls()
```
should return `_instance` directly?

xunliu replied:
`return cls()` will called `__new__()` function.

mchades commented on clients/client-python/gravitino/api/types.py:
```diff
@@ -0,0 +1,1106 @@
+
+        _instance: "BooleanType" = None
+
+        def __new__(cls):
+            if cls._instance is None:
+                cls._instance = super(Types.BooleanType, cls).__new__(cls)
+            return cls._instance
```
It appears that the code is not thread-safe.

xunliu replied:
@SophieTech88 Please add thread lock in the xxxType just like this, and need to add Test case `thread-safe`, thanks!

```
class BooleanType(PrimitiveType):
    """The boolean type in Gravitino."""

    _instance: "BooleanType" = None
    _lock = threading.Lock()

    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super(BooleanType, cls).__new__(cls)
        return cls._instance
```

xunliu replied:
Yes, I agreed @diqiu50

SophieTech88 commented:
> hi @SophieTech88 I help you improve this PR, please help me review my change parts.
> 
> 1. We can use `__new__()` implement singleton.
> 2. Added some return type and function param type.
> 3. Added some class and function comments.

@xunliu Thank you so much. It is a great help.

mchades commented on clients/client-python/tests/unittests/test_gvfs_with_local.py:
what are the codes for?

xunliu replied:
@mchades The `clients/client-python/tests/unittests/test_gvfs_with_local.py` report  waring `C0302: Too many lines in module (1102/1100) (too-many-lines)`, So I fixed it.

mchades reviewed (approved):
The type definition logic LGTM

xunliu reviewed (approved):
@SophieTech88 Thank you for your contributions.
LGTM
